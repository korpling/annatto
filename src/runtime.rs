use std::fmt::Display;

use graphannis::AnnotationGraph;

use crate::{
    error::AnnattoError,
    workflow::{StatusMessage, StatusSender},
    Result,
};

pub enum EnvVars {
    InMemory,
}

impl Display for EnvVars {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            EnvVars::InMemory => "ANNATTO_IN_MEMORY",
        };
        write!(f, "{s}")
    }
}

pub fn initialize_graph(tx: &Option<StatusSender>) -> Result<AnnotationGraph> {
    let env_var = EnvVars::InMemory.to_string();
    let on_disk = match std::env::var(env_var) {
        Ok(value) => match value.parse::<bool>() {
            Ok(v) => !v,
            Err(_) => on_error_default_storage(tx)?,
        },
        Err(_) => on_error_default_storage(&None)?, // silent, because not setting the env var is not a user error
    };
    let g = AnnotationGraph::with_default_graphstorages(on_disk)
        .map_err(|e| AnnattoError::CreateGraph(e.to_string()))?;
    Ok(g)
}

fn on_error_default_storage(tx: &Option<StatusSender>) -> Result<bool> {
    if let Some(sender) = &tx {
        sender.send(StatusMessage::Warning(format!(
            "Could not read value of environment variable {}, working on disk.",
            EnvVars::InMemory
        )))?;
    }
    Ok(true)
}

#[cfg(test)]
mod tests {
    use super::initialize_graph;
    use crate::runtime::EnvVars;

    #[test]
    fn test_storage_mode_unset() {
        // todo once graphannis exposes whether or not it's storage is disk based, let's test that here
        let tx = None;
        let g = initialize_graph(&tx);
        assert!(g.is_ok());
    }

    #[test]
    fn test_storage_mode_set_correctly() {
        // todo once graphannis exposes whether or not it's storage is disk based, let's test that here
        std::env::set_var(EnvVars::InMemory.to_string(), "true".to_string());
        let tx = None;
        let g = initialize_graph(&tx);
        assert!(g.is_ok());
    }

    #[test]
    fn test_storage_mode_set_incorrectly() {
        // todo once graphannis exposes whether or not it's storage is disk based, let's test that here
        std::env::set_var(EnvVars::InMemory.to_string(), "ehm".to_string());
        let tx = None;
        let g = initialize_graph(&tx);
        assert!(g.is_ok());
    }
}
