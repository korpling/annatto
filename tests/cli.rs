use annatto::runtime::EnvVars;
use assert_cmd::prelude::*;
use insta::assert_snapshot;
use std::{path::PathBuf, process::Command};

#[test]
fn show_help() {
    let mut cmd = Command::cargo_bin("annatto").unwrap();

    let output = cmd.arg("help").output().unwrap();
    cmd.assert().success();

    // Get output and replace version number
    let output = std::str::from_utf8(&output.stdout).unwrap();
    let output = output.replace(env!("CARGO_PKG_VERSION"), "<version>");

    assert_snapshot!(output);
}

#[test]
fn run_empty_conversion() {
    let mut cmd = Command::cargo_bin("annatto").unwrap();

    let output = cmd
        .arg("run")
        .arg("tests/data/import/empty/empty.toml")
        .output()
        .unwrap();
    cmd.assert().success();

    // Get output
    let output_err = std::str::from_utf8(&output.stderr).unwrap();
    assert_snapshot!(output_err);
    let output = std::str::from_utf8(&output.stdout).unwrap();
    assert_snapshot!(output);
}

#[test]
fn convert_to_itself() {
    // Create temporary folder for test output
    let tmp_out = tempfile::tempdir().unwrap();

    std::env::set_var("TEST_OUTPUT", tmp_out.path().to_string_lossy().as_ref());
    std::env::set_var(EnvVars::InMemory.to_string(), false.to_string());

    let mut cmd = Command::cargo_bin("annatto").unwrap();

    cmd.arg("run")
        .arg("--env")
        .arg("tests/workflows/convert_to_itself.toml")
        .output()
        .unwrap();
    cmd.assert().success();

    // Input and output files should be the same
    let original = include_str!("data/import/graphml/single_sentence.graphml");
    let converted =
        std::fs::read_to_string(tmp_out.path().join("single_sentence.graphml")).unwrap();
    pretty_assertions::assert_str_eq!(original, converted);
}

#[test]
fn run_empty_conversion_abs_path() {
    let mut cmd = Command::cargo_bin("annatto").unwrap();

    let workflow_file = PathBuf::from("tests/data/import/empty/empty.toml");

    let output = cmd
        .arg("run")
        .arg(
            workflow_file
                .canonicalize()
                .unwrap()
                .to_string_lossy()
                .as_ref(),
        )
        .output()
        .unwrap();
    cmd.assert().success();

    // Get output
    let output_err = std::str::from_utf8(&output.stderr).unwrap();
    assert_snapshot!(output_err);
    let output = std::str::from_utf8(&output.stdout).unwrap();
    assert_snapshot!(output);
}

#[test]
fn run_failing_conversion() {
    let mut cmd = Command::cargo_bin("annatto").unwrap();

    let output = cmd
        .arg("run")
        .arg("tests/data/import/failing/failing.toml")
        .output()
        .unwrap();
    cmd.assert().failure();

    // Get output
    let output_err = std::str::from_utf8(&output.stderr).unwrap();
    assert_snapshot!(output_err);
    let output = std::str::from_utf8(&output.stdout).unwrap();
    assert_snapshot!(output);
}

#[test]
fn load_complex_workflow() {
    let mut cmd = Command::cargo_bin("annatto").unwrap();

    let output = cmd
        .arg("validate")
        .arg("tests/workflows/complex_all_attributes.toml")
        .output()
        .unwrap();
    cmd.assert().success();

    // Get output
    let output_err = std::str::from_utf8(&output.stderr).unwrap();
    assert_snapshot!(output_err);
    let output = std::str::from_utf8(&output.stdout).unwrap();
    assert_snapshot!(output);
}

#[test]
fn load_complex_workflow_attr_ommited() {
    let mut cmd = Command::cargo_bin("annatto").unwrap();

    let output = cmd
        .arg("validate")
        .arg("tests/workflows/complex_some_attributes.toml")
        .output()
        .unwrap();
    cmd.assert().success();

    // Get output
    let output_err = std::str::from_utf8(&output.stderr).unwrap();
    assert_snapshot!(output_err);
    let output = std::str::from_utf8(&output.stdout).unwrap();
    assert_snapshot!(output);
}
