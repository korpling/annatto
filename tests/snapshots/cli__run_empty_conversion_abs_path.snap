---
source: tests/cli.rs
expression: output_err
---

